### ISSUE TITLE
Update fluent to 0.4.2

### ISSUE THREAD

**Issue by zbraniecki** [amethyst/amethyst#1039]
Hi Amethyst Crew!

I'm the maintainer of the `fluent-rs` crate and noticed that you're using an outdated version of our crate in your project `amethyst-locale`. We're advancing toward 1.0 and we updated the API in 0.4 which we hope we'll be able to keep toward stable release, and would appreciate if you could verify that it'll work for your project.

The update should be trivial I hope :)

**Comment by zbraniecki** [comment]
Hi!

I have no experience with Send/Sync. How hard should it be to add it? Can you direct me to any resources on that topic? Would you be able to help with a patch?

**Comment by Xaeroxe** [comment]
Here's the compile error:

```
error[E0277]: `(dyn for<'r, 's> std::ops::Fn(&'r [std::option::Option<fluent::types::FluentValue>], &'s std::collections::HashMap<std::string::String, fluent::types::FluentValue>) -> std::option::Option<fluent::types::FluentValue> + 'static)` cannot be sent between threads safely
  --> amethyst_locale\src\lib.rs:47:6
   |
47 | impl Asset for Locale {
   |      ^^^^^ `(dyn for<'r, 's> std::ops::Fn(&'r [std::option::Option<fluent::types::FluentValue>], &'s std::collections::HashMap<std::string::String, fluent::types::FluentValue>) -> std::option::Option<fluent::types::FluentValue> + 'static)` cannot be sent between threads safely
   |
   = help: the trait `std::marker::Send` is not implemented for `(dyn for<'r, 's> std::ops::Fn(&'r [std::option::Option<fluent::types::FluentValue>], &'s std::collections::HashMap<std::string::String, fluent::types::FluentValue>) -> std::option::Option<fluent::types::FluentValue> + 'static)`
   = note: required because of the requirements on the impl of `std::marker::Send` for `std::ptr::Unique<(dyn for<'r, 's> std::ops::Fn(&'r [std::option::Option<fluent::types::FluentValue>], &'s std::collections::HashMap<std::string::String, fluent::types::FluentValue>) -> std::option::Option<fluent::types::FluentValue> + 'static)>`
   = note: required because it appears within the type `std::boxed::Box<(dyn for<'r, 's> std::ops::Fn(&'r [std::option::Option<fluent::types::FluentValue>], &'s std::collections::HashMap<std::string::String, fluent::types::FluentValue>) -> std::option::Option<fluent::types::FluentValue> + 'static)>`
   = note: required because it appears within the type `fluent::entry::Entry<'static>`
   = note: required because it appears within the type `(std::string::String, fluent::entry::Entry<'static>)`
   = note: required because it appears within the type `std::marker::PhantomData<(std::string::String, fluent::entry::Entry<'static>)>`
   = note: required because it appears within the type `std::collections::hash::table::RawTable<std::string::String, fluent::entry::Entry<'static>>`
   = note: required because it appears within the type `std::collections::HashMap<std::string::String, fluent::entry::Entry<'static>>`
   = note: required because it appears within the type `fluent::bundle::FluentBundle<'static>`
   = note: required because it appears within the type `Locale`

error: aborting due to 4 previous errors

For more information about this error, try `rustc --explain E0277`.
error: Could not compile `amethyst_locale`.
warning: build failed, waiting for other jobs to finish...
error: build failed
```

Send/Sync is usually implemented automatically, but some API choices can prevent this from happening.  Here's some more info: https://doc.rust-lang.org/book/second-edition/ch16-04-extensible-concurrency-sync-and-send.html

If I find myself with some extra time I'll be happy to shoot a patch to you, but it also would probably just be faster to not bother waiting for me.

**Comment by Xaeroxe** [comment]
Here's more documentation on the traits:

https://doc.rust-lang.org/beta/nomicon/send-and-sync.html

**Comment by zbraniecki** [comment]
Okay... that seems non-trivial then! Glad I asked :)

So, if I understand correctly, the issue comes from https://github.com/projectfluent/fluent-rs/blob/master/fluent/src/entry.rs#L7-L8

We use that to allow developers to define built-in functions that localizers can use in their translations.

Here are some intro docs: https://projectfluent.org/fluent/guide/functions.html

and here's an example of an app using it: https://github.com/projectfluent/fluent-rs/blob/master/fluent/examples/functions.rs

Does it sound to you like this sort of code has to be single-threaded by nature and if we want to support multi-threaded code we cannot allow for functions to be registered on `FluentBundle`?

**Comment by Xaeroxe** [comment]
So I did a little bit of research and basically here's the problem as I see it:

[this](https://docs.rs/fluent/0.4.2/fluent/entry/enum.Entry.html#variant.Function) variant contains a Box with a trait object.  Rust is not currently guaranteeing that trait object is also Send + Sync though so that explodes into the whole enum being not Send + Sync, which then escalates into the bundle type.

Simplest solution for our use case would probably be to just add `+ Send + Sync` to the trait object in the box, but this could produce problems for you guys later if someone wants to provide a function that doesn't impl `Send + Sync`.  On the other hand, it's kind of rare to make a closure that isn't `Send + Sync`.

I'm not super sure what the best way to resolve this is.  Here's some options.

1. Just add `Send + Sync` to it and say "too bad so sad" to the (small?) set of users who really didn't want that.
2. Add a feature to the crate which alters this type to require `Send + Sync`.  Serves both use cases but might arguably be a tad clunky?
3. Provide two `Entry` enums, where one is Send + Sync and the other isn't and let the user pick which one to use.  Might cause unnecessary maintenance work for you guys though.

**Comment by zbraniecki** [comment]
I'm totally in favor of (1). The closures should be rather trivial ones and I hope they can always be Send/Sync. If not, we'll at least be guarded by a backward incompatible change that we'll have to justify rather than a silent breakage as we have today.

**Comment by zbraniecki** [comment]
Got the entry.rs to have `+ Sync + Send` but now I need to also add it to bundles.rs and this produces an error:

```rust
    pub fn add_function<F>(&mut self, id: &str, func: F) -> Result<(), FluentError>
    where
        F: 'bundle
            + Fn(&[Option<FluentValue>], &HashMap<String, FluentValue>) + Send + Sync -> Option<FluentValue>,
    {
```

```
▶ cargo check
    Checking fluent v0.4.2 (/home/zbraniecki/projects/fluent/fluent-rs/fluent)
error: expected one of `(`, `+`, `,`, `::`, `<`, or `{`, found `->`  
  --> fluent/src/bundle.rs:86:87                                     
   |                                                                 
86 |             + Fn(&[Option<FluentValue>], &HashMap<String, FluentValue>) + Send + Sync -> Option<FluentValue>,
   |                                                                                       ^^ expected one of `(`, `+`, `,`, `::`, `<`, or `{` here
                                                                     
error: expected one of `(`, `+`, `,`, `::`, `<`, `async`, `const`, `crate`, `default`, `existential`, `extern`, `fn`, `pub`, `type`, `unsafe`, or `}`, found `->`
  --> fluent/src/bundle.rs:86:87                                     
   |                                                                 
86 |             + Fn(&[Option<FluentValue>], &HashMap<String, FluentValue>) + Send + Sync -> Option<FluentValue>,
   |                                                                                       ^^ expected one of 16 possible tokens here
                                                                     
error: aborting due to 2 previous errors                             
                                                                     
error: Could not compile `fluent`.
```

Do you know why?

**Comment by Xaeroxe** [comment]
Yeah, it looks like you need to add it after the return type.  You've split up your function's signature declaration

**Comment by Xaeroxe** [comment]
Re-opening, this issue should be closed by #1040
